clean_user_registry_secrets/clean_user_registry_secrets.py

```diff
@@ -0,0 +1,151 @@
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Scripts used to remove user registry secrets in k8s"""
+
+from datetime import datetime, timedelta
+from pathlib import Path
+import argparse
+import re
+import logging
+from functools import reduce
+
+from kubernetes import client
+from kubernetes.config.incluster_config import (
+    SERVICE_CERT_FILENAME,
+    SERVICE_TOKEN_FILENAME,
+    InClusterConfigLoader,
+)
+
+
+def find_pod_by_secret(secret, k8s_client):
+    """Find the user jupyterhub podname based on the registry pull secret."""
+    label_keys = ['renku.io/commit-sha', 'renku.io/projectName', 'renku.io/username']
+    label_selector = []
+    for label_key in label_keys:
+        label_selector.append(f"{label_key}={secret.metadata.labels[label_key]}")
+    label_selector = ",".join(label_selector)
+
+    pod_list = k8s_client.list_namespaced_pod(
+        secret.metadata.namespace,
+        label_selector=label_selector,
+    )
+    if len(pod_list.items) > 0:
+        return pod_list.items[0].metadata.name
+    return None
+
+
+def remove_user_registry_secret(namespace, k8s_client, min_secret_age_hrs=0.25):
+    """Used in a cronjob to periodically remove old user registry secrets"""
+    secret_name_regex = ".+-registry-[a-z0-9-]{36}$"
+    label_keys = ['renku.io/commit-sha', 'renku.io/projectName', 'renku.io/username']
+    logging.info(
+        f"Checking for user registry secrets whose "
+        f"names match the regex: {secret_name_regex}"
+    )
+    secret_list = k8s_client.list_namespaced_secret(namespace)
+    for secret in secret_list.items:
+        # loop through secrets and find ones that match the predefined regex
+        secret_name = secret.metadata.name
+        secret_name_match = re.match(secret_name_regex, secret_name)
+        # calculate secret age
+        tz = secret.metadata.creation_timestamp.tzinfo
+        secret_age = datetime.now(tz=tz) - secret.metadata.creation_timestamp
+        min_secret_age = timedelta(hours=min_secret_age_hrs)
+        if (
+            secret_name_match is not None
+            and secret.type == "kubernetes.io/dockerconfigjson"
+            and secret.metadata.labels.get("component") == "singleuser-server"
+            and reduce(  # check that label keys for sha, project and username are present
```

Comment:
maybe

```
all(label_key in secret.metadata.labels for label_key in label_keys)
```

is easier to understand? I'm pretty sure it would do the same thing but I could be wrong.